ratis-server/src/main/java/org/apache/ratis/server/impl/RaftServerImpl.java

```diff
@@ -593,6 +593,30 @@ public boolean isLeaderReady() {
     return true;
   }
 
+  /**
+   * Return status of the RaftServer.
+   * @return LeaderStatus
+   */
+  public ServerStatus getServerStatus() {
```

Comment:
Do we really need this?  This method is almost the same as isLeaderReady().

Reply:
There is a slight difference, isLeaderReady will return false when it is leader but not ready, this method helps in finding that.

isLeaderReady -> True when leader and ready
                          -> false when not leader
                          -> false when leader and not ready

This API can help can let the caller know exactly the server status.

Reply:
One reason is in OM HA we want to check isLeader and isLeaderReady and these are implemented as below.


  public boolean isLeader() {
    Preconditions.checkState(server instanceof RaftServerProxy);
    RaftServerImpl serverImpl = null;
    try {
      serverImpl = ((RaftServerProxy) server).getImpl(raftGroupId);
      if (serverImpl != null) {
        return serverImpl.isLeader();
      }
    } catch (IOException ioe) {
      // In this case we return not a leader.
      LOG.error("Fail to get RaftServer impl and therefore it's not clear " +
          "whether it's leader. ", ioe);
    }
    return false;
  }

  /**
   * Return true, if the current OM node is leader and in ready state to
   * process the requests.
   * @return
   */
  public boolean isLeaderReady() {
    Preconditions.checkState(server instanceof RaftServerProxy);
    RaftServerImpl serverImpl = null;
    try {
      serverImpl = ((RaftServerProxy) server).getImpl(raftGroupId);
      if (serverImpl != null) {
        return serverImpl.isLeaderReady();
      }
    } catch (IOException ioe) {
      // In this case we return not a leader.
      LOG.error("Fail to get RaftServer impl and therefore it's not clear " +
          "whether it's leader. ", ioe);
    }
    return false;
  }

With this API i can call new API and return error codes to the client accordingly for retry.

Reply:
We already have isLeader() and isLeaderReady() covering all the cases.  Could you use these two methods for now?  

RaftServerImpl is an internal Ratis class but not a public API.  Ozone should not use it directly.  If there is a need, we should define a public API but not adding methods to RaftServerImpl.

Reply:
BTW, RaftServerProxy is also an internal Ratis class.

Reply:
Because in OM HA, we want to check isLeaderReady and then start serving read requests. 
https://issues.apache.org/jira/browse/HDDS-4484

The main reason was not to call multiple time getImpl(groupID) and then find isLeader and isLeaderReady.

We use RaftServer Builder build() to construct and use that object to make calls to Ratis.


```
    this.server = RaftServer.newBuilder()
        .setServerId(this.raftPeerId)
        .setGroup(this.raftGroup)
        .setProperties(serverProperties)
        .setStateMachine(omStateMachine)
        .build();


```
Let me know what would be the best way to expose this information from Ratis? By adding it to RaftServer?

https://issues.apache.org/jira/browse/HDDS-4484 Jira in Ozone for OM HA.

Reply:
Could you use the following code in Ozone for the moment?
```
      final RaftServerImpl impl = ((RaftServerProxy)raftServer).getImpl(groupId);
      if (!impl.isLeader()) {
        // not leader
      } else if (impl.isLeaderReady()) {
        // leader and ready
      } else {
        // leader but not ready
      }
```
 I will work on creating new APIs.